### PR TITLE
Code-First & Type-First Federation 2 Support

### PR DESCRIPTION
**Pull request by Shane32**
Based on #3144 by @Mithras

- New code is in namespace GraphQL.Federation
- Old code in namespace GraphQL.Utilities.Federation has been marked as obsolete
- Federation code is unified into `.AddFederation()` for both schema-first, code-first and type-first use cases.
- Old schema-first syntax is marked as obsolete, but will function without any (or very few) changes
- Code first supported, mostly by @Mithras 
- Type first supported
- Apollo Router now runs an actual test against four live subgraph schemas (aka demo apps) to form a single supergraph:
  - Schema first, new syntax, federation 1 compatible
  - Schema first, old syntax, federation 1 compatible
  - Code first, federation 2 compatible
  - Type first, federation 2 compatible

Breaking changes:
- A couple classes were moved.
- The `__typename` field is not implicitly added to federation requests

Prerequisites:
- [x] #3952 
- [x] #3949 
- [x] #3951
- [x] #3965 
- [x] #3975

Split into separate PRs:
- [x] #3957 
- [x] #3958 
- [x] #3961 
- [x] #3960
- [x] #3963 
- [x] #3964 
- [x] Add sample applications / tests -- this PR

To-do / nice-to-have:
- [ ] Analyzer rule for new `[FederationResolver]` attribute to ensure it is used correctly
- [ ] #3962 
- [x] New tests for Sample3 and Sample4 web applications
- [x] Additional federation testing for Sample3 and Sample4
- [x] Move new extension methods to GraphQL.Federation namespace rather than GraphQL.Federation.Extensions
- [x] Add documentation
- [x] Break PR into smaller pieces and merge each piece separately
- [x] Merge master/develop into PR
- [x] Add xml comments to all federation classes/etc

Moved out of scope:
- [ ] New validation rule - see https://github.com/graphql/graphql-spec/pull/1098
- [ ] Consider: auto-hide any unreferenced types in the schema (??)
- [ ] Consider: post-initialization schema visitor

### CODE REVIEW

File: src/GraphQL.sln
Comment (Shane32) [inline src/GraphQL.sln:135]:
```suggestion
```

File: src/GraphQL.sln
Comment (Shane32) [inline src/GraphQL.sln:262]:
```suggestion
```

**Comment by Shane32** [comment]
@gao-artur How involved do you want to be with this PR?

So far I've taken what others have done before me and made the following modifications:
- `ToObject` design was changed for v8, requiring the necessary functionality rewritten for federation
- Rewrite `IFederationResolver` to return `ValueTask<object?>` and properly support asynchronous tasks
- Fix entity resolver to support serial and parallel execution strategies
- Repair tests and other misc for v8 compatibility
- Remove built-in camel-case conversion of field names within attributes (this means `nameof` won't work, but now it is compatible with fields of any case, as camel-case is not mandated by graphql)
- Utilize new schema printer, fix where directives are stored
- Fix extension methods for v8 to use `IMetadataWriter` and similar (new for v8)
- Remove `NeverType` from codebase (requires #3927 )

Note: I read a note in the code that types' permissions are not validated from the federation resolver.  This is true for any union type; the union type itself is validated but not the specific type returned, as it is not known until the resolver executes.  We should consider fixing this for v8, not just for federation, but for unions or interfaces in general (or else be sure to add the applicable notes to the documentation that this occurs).

**Comment by Shane32** [comment]
cc: @Mithras

**Comment by gao-artur** [comment]
Hi @Shane32, sorry I was a bit off last month. Changes at work, and I hardly have free time now.
I'm not sure I have enough knowledge of the GraphQL.NET internals to review this PR, but I definitely want to know how it works because I want to adopt the Federation shortly after it is released. So, I'll review all your PR's anyway, even after you merge them.
And I'll try to finish the input graph type analyzers during this month because I think they must be ready for v8.0.

**Comment by Shane32** [comment]
@Mithras @michael-watson @mac2000  I'm going to try to put time into this PR over the next couple weekends, so GraphQL.NET v8 can be released.  It's at the point now where I think the code is really close to what it should be.  Let me know if you'd like to spend time reviewing it or assisting also.

**Comment by Shane32** [comment]
@joemcbride I've been working on GraphQL.NET Federation support for code-first and "type-first" approaches in addition to the schema-first approach already available, basing my efforts on prior work done by @Mithras and others.  I've also unified the schema-first setup with the other approaches, while still supporting (for v8) any legacy schema-first code (with applicable code marked as obsolete).  Let me know if you'd like to review the changes here.  I know there's a lot but if you're interested, I can guide you what will affect schema-first.  Mainly I'd look at these four subgraph projects, which each are part of a single supergraph (tested together within CI):
- GraphQL.Federation.SchemaFirst.Sample1 - schema-first with new syntax
- GraphQL.Federation.SchemaFirst.Sample2 - schema-first with old syntax
- GraphQL.Federation.CodeFirst.Sample3 - code-first
- GraphQL.Federation.TypeFirst.Sample4 - "type-first"

@sungam3r @gao-artur Let me know if you'd like to participate in this PR also.

I'd really like to get GraphQL.NET v8 released with code-first federation support, plus the other new enhancements that are ready to be released.

**Comment by mac2000** [comment]
@Shane32 in our case we have bazillion of services running on graphql-net with self made federation introduced back then in issue which I can not find anymore

side by side with new services based on Hotchocolate

I will ask our backend engineer to see if he will be able to get up and running one of the services based on this branch and come back with feedback if everything workedout

**Comment by gao-artur** [comment]
I have yet to learn the theory, but I'm trying to understand how the federated schemas should work based on the sample projects. I see that the `Sample3` project defines the `UserGraphType` with two fields: `id` and `reviews`. The `Sample4` project defines the `User` type with fields `id` and `username`. So, if I start and configure the Apollo client, I will see the `user` type with 3 fields: `id`, `username` and `reviews`? Is that correct?
Btw, why the Sample3 and 4 projects are not in the solution file?

**Comment by Shane32** [comment]
> Btw, why the Sample3 and 4 projects are not in the solution file?

Weird.  Fixed.

> So, if I start and configure the Apollo client, I will see the user type with 3 fields: id, username and reviews? Is that correct?

Yes, that is correct.  Conceptually, the samples are microservices as follows:

1. Categories microservice.  Has a Category type but doesn't depend on any other microservices.
2. Products microservice.  Has a Product type, and extends Category to view all products in that category.
3. Reviews microservice.  Has a Review type, and extends Product and User types.
4. Users microservice.  Has a User type but doesn't depend on any other microservices.

See `federation-request-4.json` and `federation-response-4.json` for a sample query that runs a query across all services, where the Apollo Router will need to wait for an answer from certain subgraphs to be able to forward additional requests to other subgraphs, until it can assemble the entire response.

**Comment by gao-artur** [comment]
So, if I want to resolve `username` from `review.author`, Apollo will send an HTTP request with the user id to the Users microservice? What if there is a list of `users`? Does it support DataLoader to issue a single HTTP request?

**Comment by Shane32** [comment]
> So, if I want to resolve `username` from `review.author`, Apollo will send an HTTP request with the user id to the Users microservice? What if there is a list of `users`? Does it support DataLoader to issue a single HTTP request?

Yes, Apollo will send a list of entities that need resolving together in a single HTTP request.  Within GraphQL.NET we can use data loaders to issue a single request to the underlying database.

**Comment by gao-artur** [comment]
Thanks! I'll try to read more about Federation on the weekend and try to help with the review.

**Comment by gao-artur** [comment]
I'm still learning federation basics, but I noticed the federation subgraph can define a schema with [types only](https://www.apollographql.com/docs/federation/federated-types/overview/#subgraph-schemas) (`Reviews` schema) and without any query field. Will GraphQL.NET allow such schemas when `.AddFederation()` is used?

**Comment by Shane32** [comment]
It should.  Technically there should still be a minimum of one field, `_services`, which is automatically added to the query type.  `AddFederation` should add this before running schema validation checks.  We should add a test for this scenario, however.

Chances are that `AddFederation` does not yet handle the scenario where if there are zero resolvable types, then the `_entity` field is not added to the query type.  I should add a test for that as well.

**Comment by Shane32** [comment]
> I'm still learning federation basics, but I noticed the federation subgraph can define a schema with [types only](https://www.apollographql.com/docs/federation/federated-types/overview/#subgraph-schemas) (`Reviews` schema) and without any query field. Will GraphQL.NET allow such schemas when `.AddFederation()` is used?

I've revised Sample3 (the 'Reviews' schema) to remove the field added to the Query type.  I've also added support for subgraphs with no resolvable entities.  For those subgraphs, the `_Entity` type and `_entities` field is removed from the schema.  Finally, I added additional tests for Sample3 and Sample4.

File: src/GraphQL/Federation/Extensions/FederationHelper.cs
Comment (gao-artur) [inline src/GraphQL/Federation/Extensions/FederationHelper.cs:154]: Is there a plan to support higher [specs](https://www.apollographql.com/docs/federation/federation-versions/) versions?
From what I understand, higher specs require clients to upgrade their router version. So, maybe this version can be set dynamically based on the features used. For example, if the `@interfaceObject` directive was applied during the schema definition, the version here should be updated to v2.3.

**Comment by Shane32** [comment]
@vadymkutsenko Thanks for testing - it's very encouraging!

> app.UseGraphQL doesn't work

Makes sense - probably because the GraphQL.Server.Transports.AspNetCore package referenced ISchema inside the NuGet GraphQL.dll and you have compiled your own ISchema.  Chances are that you can solve this by using the beta build from GitHub Packages, if you so desired to do so.  This will all be cleaned up when v8 is released of course.

File: src/GraphQL/Federation/Extensions/FederationHelper.cs
Comment (Shane32) [reply src/GraphQL/Federation/Extensions/FederationHelper.cs:154]: I think I'll change the `AddFederation` method argument to be a builder method, so we can add features or configuration settings going forwards.  Perhaps the federation version should be a string property?

Comment (gao-artur) [reply src/GraphQL/Federation/Extensions/FederationHelper.cs:154]: Yep, I thought about such config as well. It can be a "higher supported spec version" config. Then, we can add a validation that will throw if the user used a feature from a higher version. So, I can see two approaches here:
1. Automatic detection of the minimal required version based on used features
2. Explicit higher version set by configuration

I think both are useful.

File: src/GraphQL.Tests/Utilities/FederatedSchemaBuilderTests.cs
Comment (Shane32) [inline src/GraphQL.Tests/Utilities/FederatedSchemaBuilderTests.cs:70]: `__typename` is no longer automatically injected into the query

File: src/GraphQL.Tests/Utilities/FederatedSchemaBuilderTests.cs
Comment (Shane32) [inline src/GraphQL.Tests/Utilities/FederatedSchemaBuilderTests.cs:91]: `__typename` is no longer automatically injected into the query.

File: src/GraphQL.Tests/Utilities/FederatedSchemaBuilderTests.cs
Comment (Shane32) [inline src/GraphQL.Tests/Utilities/FederatedSchemaBuilderTests.cs:217]: `__typename` is no longer automatically injected into the query.

File: src/GraphQL/Federation/Types/LinkImportGraphType.cs
Comment (gao-artur) [inline src/GraphQL/Federation/Types/LinkImportGraphType.cs:8]: Why does it derive from Any and not String?

**Comment by Shane32** [comment]
@gao-artur This PR now only contains additional tests, some created by me, others by @Mithras .

File: src/GraphQL/Federation/Extensions/FederationHelper.cs
Comment (Shane32) [reply src/GraphQL/Federation/Extensions/FederationHelper.cs]: > Yep, I thought about such config as well. It can be a "higher supported spec version" config. Then, we can add a validation that will throw if the user used a feature from a higher version. So, I can see two approaches here:
> 
> 1. Automatic detection of the minimal required version based on used features
> 2. Explicit higher version set by configuration
> 
> I think both are useful.

Completed

**Comment by gao-artur** [comment]
This is incredible work! Thank you and @Mithras for making this possible!

**Review by gao-artur — APPROVED** [review]
This is incredible work! Thank you and @Mithras for making this possible!

**Pull request merged**
